src/libraries/System.Private.DataContractSerialization/src/System/Xml/XmlBinaryReader.cs

```diff
@@ -1350,14 +1350,14 @@ private unsafe int ReadArray(float[] array, int offset, int count)
 
         public override int ReadArray(string localName, string namespaceUri, float[] array, int offset, int count)
         {
-            if (IsStartArray(localName, namespaceUri, XmlBinaryNodeType.FloatTextWithEndElement))
+            if (IsStartArray(localName, namespaceUri, XmlBinaryNodeType.FloatTextWithEndElement) && BitConverter.IsLittleEndian)
```

Comment:
I think it'd be better for the BitConverter.IsLittleEndian check to be first in all of these cases, e.g.
```C#
if (BitConverter.IsLittleEndian && IsStartArray(localName, namespaceUri, XmlBinaryNodeType.FloatTextWithEndElement))
```
If `IsStartArray` isn't inlineable, the JIT needs to assume it might have side effects that can't be elided, and so to preserve order of operations, in the big endian case the JIT would need to compile this as:
```C#
IsStartArray(localName, namespaceUri, XmlBinaryNodeType.FloatTextWithEndElement);
return base.ReadArray(localName, namespaceUri, array, offset, count);
```
But with the check first, when it's false the second clause would never be evaluated, so the JIT can compile it just as:
```C#
return base.ReadArray(localName, namespaceUri, array, offset, count);
```

Reply:
I believe the author was just following the pattern we had established on other types like short/int/long where this check is done second. I don't feel strongly about this, but my inclination would be to leave the order as is for 7.0 and we can invert the check for all types in 8.0 where we have more runway. If you feel strongly about this though, I wouldn't object to making the change here.

Reply:
We should fix it in main. I don't feel strongly about it for release/7.0, since to my knowledge we don't actually ship supported S390X bits from that branch.  I am curious, though, where we have existing BitConverter.IsLittleEndian checks at the end of a clause; those should be fixed, too.